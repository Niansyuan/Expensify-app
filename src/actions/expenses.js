import { child, get, push, ref, remove, update } from '@firebase/database';
import database from '../firebase/firebase';

//ADD_EXPENSES
// const action=(({}={})=>({}))
export const addExpenses = (expense) => ({
    type: 'ADD_EXPENSES',
    expense
});

//dispatch addExpenses to change store (add this action after introduce firebase)
export const startAddExpenses = (expenseDate = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createAt = 0
        } = expenseDate;
        const expense = { description, note, amount, createAt }
        const rootRef = ref(database, `users/${uid}/expenses`)
        return push(rootRef, expense).then((rootRef) => {
            dispatch(addExpenses({
                id: rootRef.key,
                ...expense
            }));
        });
    };
};

//REMOVE_EXPENSES
export const removeExpenses = ({ id }) => ({
    type: 'REMOVE_EXPENSES',
    id
})

export const startRemoveExpenses = ({ id }) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return remove(ref(database, `users/${uid}/expenses/${id}`)).then(() => {
            dispatch(removeExpenses({ id }))
        });
    };
};

//EDIT_EXPENSES
export const editExpenses = (id, updates) => ({
    type: 'EDIT_EXPENSES',
    id,
    updates
});

export const startEditExpenses = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return update(ref(database, `users/${uid}/expenses/${id}`), updates).then(() => {
            dispatch(editExpenses(id, updates))
        });
    };
};


//SET_EXPENSES (add this action after introduce firebase)
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//startSetExpenses (add this action after introduce firebase)
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        //fetch all expenseData once
        return get(child(ref(database), `users/${uid}/expenses`)).then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });
            dispatch(setExpenses(expenses));
        });
    };
};