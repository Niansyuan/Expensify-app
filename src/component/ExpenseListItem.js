import React from "react";
import {Link} from 'react-router-dom';
import moment from "moment";
import numeral from "numeral";

const ExpenseListItem=({id,description,amount,createAt})=>{
    return (
        <div>
            <Link to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>
                NT{numeral(amount).format('$0,0.00')}
                __
                {moment(createAt).format('MMM Do, YYYY')}
            </p> 
        </div>
    )
}

export default ExpenseListItem;