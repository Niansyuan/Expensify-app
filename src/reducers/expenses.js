//expenses reducer
const expensesReducerDefaultState=[]
const expensesReducer=(state=expensesReducerDefaultState,action)=>{
    switch(action.type){
        case 'ADD_EXPENSES':
            return [...state,action.expense]; //等於state.concat(action.expense)
        case 'REMOVE_EXPENSES':
            return state.filter(({id})=>{
                return id!== action.id
            });
        case 'EDIT_EXPENSES':
            return state.map((expense)=>{
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                }else{
                    return expense
                }
            });
        default:
            return state;
    }
}
export default expensesReducer;