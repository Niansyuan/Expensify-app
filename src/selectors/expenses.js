import moment from "moment";

//Get visible expenses
export default(expenses,{text,sortBy,startDate,endDate})=>{
    return expenses.filter((expense)=>{
        //all three are ture,we want to return true from filter
        const createAtMoment=moment(expense.createAt);
        const startDateMatch= startDate ? startDate.isSameOrBefore(createAtMoment,'day'):true;
        const endDateMatch= endDate ? endDate.isSameOrAfter(createAtMoment,'day'):true;
        const textMatch= expense.description.toLowerCase().includes((text.toLowerCase()));
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createAt < b.createAt? 1:-1
        };
        if(sortBy === 'amount'){
            return a.amount < b.amount? 1:-1
        };
    })
}