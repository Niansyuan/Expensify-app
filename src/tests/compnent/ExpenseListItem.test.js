import React from "react";
import ExpenseListItem from "../../component/ExpenseListItem";
import { shallow } from "enzyme";
import expenses from "../fixture/expenses";
import "../setupTests";

test('should render expenseListItem with fixture data',()=>{
    const wrapper=shallow(<ExpenseListItem {...expenses[0]}/>)
    expect(wrapper).toMatchSnapshot();
})