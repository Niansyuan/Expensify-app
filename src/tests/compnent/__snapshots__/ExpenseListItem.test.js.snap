// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenseListItem with fixture data 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    NT
    $100.00
    __
    Jan 1st, 1970
  </p>
</div>
`;
